server/src/main/java/org/jboss/as/server/controller/git/ElytronClientSshdSessionFactory.java

```diff
@@ -0,0 +1,236 @@
+
+    private static AuthenticationContextConfigurationClient CLIENT = AccessController.doPrivileged(AuthenticationContextConfigurationClient.ACTION);
+    private final AuthenticationContext context;
+
+    private final Map<Tuple, Iterable<KeyPair>> defaultKeys;
+    private String[] defaultIdentities;
+    private KeyPair keyPair;
+    private URI uri;
+    private String knownHostsFile;
+
+    public ElytronClientSshdSessionFactory(URI authenticationConfig) throws ConfigXMLParseException, GeneralSecurityException {
+        if (authenticationConfig != null) {
+            context = ElytronXmlParser.parseAuthenticationClientConfiguration(authenticationConfig).create();
+        } else {
+            context = null;
+        }
+
+        this.defaultKeys = new ConcurrentHashMap<>();
+        this.setSshDirectory(SSHCredential.DEFAULT_SSH_DIRECTORY);
+        this.defaultIdentities = SSHCredential.DEFAULT_PRIVATE_KEYS;
+        this.keyPair = null;
+        this.knownHostsFile = SSHCredential.DEFAULT_KNOWN_HOSTS;
+    }
+
+    @Override
+    @NonNull
+    protected List<Path> getDefaultIdentities(@NonNull File sshDir) {
+        return Stream.of(defaultIdentities)
+                .map(s -> sshDir.toPath().resolve(s))
+                .filter(path -> Files.exists(path))
+                .collect(Collectors.toList());
+    }
+
+    @Override
+    public SshdSession getSession(URIish urIish, CredentialsProvider credentialsProvider, FS fs, int tms) throws TransportException {
+        if (context == null) return super.getSession(urIish, credentialsProvider, fs, tms);
+        try {
+            this.uri = urIishToUri(urIish);
+            AuthenticationConfiguration config = CLIENT.getAuthenticationConfiguration(this.uri, context);
+            CredentialCallback locationCallback = new CredentialCallback(SSHCredential.class, null);
+            CredentialCallback keyPairCallback = new CredentialCallback(KeyPairCredential.class, null);
+            Callback[] callbacks = {locationCallback, keyPairCallback};
+            CLIENT.getCallbackHandler(config).handle(callbacks);
+
+            if (locationCallback.getCredential() != null) {
+                SSHCredential sshCredential = locationCallback.getCredential().castAs(SSHCredential.class);
+                this.setSshDirectory(sshCredential.getSshDirectory());
+                this.defaultIdentities = sshCredential.getPrivateKeyIdentities();
+                this.knownHostsFile = sshCredential.getKnownHostsFile();
+            }
+
+            if (keyPairCallback.getCredential() != null) {
+                KeyPairCredential keyPairCredential = keyPairCallback.getCredential().castAs(KeyPairCredential.class);
+                this.keyPair = keyPairCredential.getKeyPair();
+            }
+
+        } catch (IOException  | UnsupportedCallbackException | URISyntaxException ex) {
+            ServerLogger.ROOT_LOGGER.failedToLoadSSHCredentials(ex, ex.getMessage());
```

Comment:
Should this throw an exception or is logging it fine?

Reply:
Ah yeah I've updated it to throw an exception, thanks!